proposals/0000-extension-lifecycle-framework.md

```diff
@@ -0,0 +1,108 @@
+ 
+ * `Mature` extensions are considered to be finished and are not expected to undergo regular changes. These features can be used without worry of unexpected changes, and they are not known to contain serious design or implementation deficiencies. Any breaking change to a mature extension will be announced well in advance of the change being made, with a migration path provided if possible. Some extensions in this category might be `MultiWayIf`, `MonoLocalBinds`, and `ViewPatterns`.
+
+ * `Deprecated` extensions are considered to be design or implementation dead ends, and should probably not be used in new code. Deprecating an extension _must_ come with a statement about the longevity of the extension. This need not be a commitment to remove it at a concrete time, but is a commitment to either remove it or transition it to another state at a time stated by GHC HQ when it is marked as `Deprecated`. Any extension will be deprecated prior to removal. Deprecated extensions are expected to include `IncoherentInstances`, `OverlappingInstances`, `Rank2Types`, `RecordPuns`, and `TypeInType`.
+ 
+ * `Legacy` extensions are explicitly not recommended for new code, much like `Deprecated`, however, they are expected to be supported indefinitely. This may be to maintain an older language set, maintain a bridge to a newer feature, or otherwise keep backward compatibility. Expected legacy extensions include `CUSKs`, `DeepSubsumption`,  and `NPlusKPatterns`.
+
+Above we list expected categorizations of several extensions based on statements in the user's guide. We do not attempt to prescribe any particular categorizations as part of this proposal and simply provide a small number of expectations from the perspective of someone reading the existing documentation.
+
+The expected extension lifecycle includes the following transitions:
+ * `Experimental` -> `Mature`
+ * `Experimental` -> `Deprecated`
+ * `Mature` -> `Legacy`
+ * `Mature` -> `Deprecated`
+ * `Legacy` -> `Deprecated`
+ * `Deprecated` -> `Legacy`
+
+However, it also seems plausible that new knowledge might from time to time cause a mature extension to once again be considered experimental, e.g. in the face of soundness bugs or subtle interactions with other features. We also do not rule out any transition explicity to allow for unforeseen circumstances. 
+
+For existing, or future, language sets such as `GHC2021` or `Haskell98`, it is expected that none of the contained extensions would be `Experimental`.
+
+### 2.2. User Interface
+
+Haskell users have different tolerances for risks related to language change. Some derive great value from using the newest features, while others are conservative. GHC should serve the whole spectrum.
+
+There will be an additional set of warnings:
+ * `-WXDeprecated`: Issue a warning when a deprecated extension is used
+ * `-WXMature`: Issue a warning when a mature extension is used
```

Comment:
Not on its own. It seems plausible that it might be useful together with some more specific flags to configure a policy, but it's mostly there for the sake of consistency and completeness.

Reply:
Assuming it means "when a mature extension is mentioned in a `LANGUAGE` pragma" then I can imagine it being useful for identifying modules with `LANGUAGE` pragmas that are redundant given the global settings.

Reply:
This is not what it means - but I think we addressed this later on in the discussion.